### PR TITLE
Refactor: Use Object ID type rather than long type in APIs [tp-tests]

### PR DESCRIPTION
**Pull request by li-boxuan**
This commit refactors the codebase by changing all long-type ID into Object-type in
all method signatures, return types, etc. No functionality is affected, except that
IDs are now stored as objects rather than primitives in the memory, which increases
memory footprint by a little bit. This is the first step to implement #1221.

This PR is to replace https://github.com/JanusGraph/janusgraph/pull/2942 because the benchmark PR commenter is only available from the JanusGraph repo.

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

**Comment by li-boxuan** [comment]
Just did a force-commit to force benchmark CI to run again. The result [here](https://github.com/JanusGraph/janusgraph/pull/3162#commitcomment-79831336) was too good to be true, and was likely caused by variance in GitHub Actions environment itself.

**Comment by li-boxuan** [comment]
It seems the GitHub CI environment varies a lot these days - thus, the benchmark result is not very reliable. I ran benchmarks locally and seems there's no regression or improvement by this PR.

FYI the benchmark result is: [benchmark.md](https://github.com/JanusGraph/janusgraph/files/9234044/benchmark.md)

**Comment by analytically** [comment]
@li-boxuan can rebase?

File: janusgraph-test/src/test/java/org/janusgraph/graphdb/query/vertex/VertexIdListTest.java
Comment (FlorianHockmann) [inline janusgraph-test/src/test/java/org/janusgraph/graphdb/query/vertex/VertexIdListTest.java:81]: Shouldn't this be:
```suggestion
        VertexIdList list = new VertexIdList(tx, new ArrayList<>(Arrays.asList("x1", "x2", "y1")), false);
```

**Review by FlorianHockmann — COMMENTED** [review]
Sorry for the extremely late review, I had actually started reviewing this PR months ago and also wrote some first inline comments back then, but didn't finish the review and then never really got back to it.
Any way, great contribution. The changes look good to me and I only wrote minor comments, but I'm not sure whether other places also need to be changed. But I guess, if it builds and all tests pass, then it should be OK overall.

File: janusgraph-test/src/test/java/org/janusgraph/graphdb/query/vertex/VertexIdListTest.java
Comment (li-boxuan) [reply janusgraph-test/src/test/java/org/janusgraph/graphdb/query/vertex/VertexIdListTest.java:81]: Nope. This is to showcase that merging a sorted list with an unsorted list would result in an unsorted list.

Comment (FlorianHockmann) [reply janusgraph-test/src/test/java/org/janusgraph/graphdb/query/vertex/VertexIdListTest.java:81]: Understood, the naming scheme just made me expect this to be basically the same as `testMergeUnsortedLongIdLists`, just with a mixture of strings and longs. You could maybe add this as an additional test specifically for the scenario of merging a sorted with an unsorted list to reduce this confusion, but it's also not really important.

**Review by FlorianHockmann — APPROVED** [review]

File: janusgraph-driver/src/main/java/org/janusgraph/graphdb/relations/RelationIdentifier.java
Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/graphdb/relations/RelationIdentifier.java:63]: We cannot. A bit context here: a `RelationIdentifier` is essentially an id for each edge / property, which consists of `outVertexId`, `typeId`, `relationId` (a unique long value id for this edge/property), and an optional `inVertexId` (only existent for edges but not properties).

Previously, the `inVertexId` is initialized as 0. On [RelationIdentifier.java line 92 before this change](https://github.com/JanusGraph/janusgraph/pull/3162/files#diff-4181397444bf720764051d1d50a66b0846fe2e5c2f0279d147c5ad40c2698ab5L92), it guarantees this `getInVertexId` is only called when `inVertexId != 0`. Now we essentially move this check to the caller side (see IndexRecordUtil.java:266 in the new code, which doesn't really make much of a difference), so we have to remove the `!=  0`, or, in the new codebase, `!= null` check from this function.

Comment (FlorianHockmann) [reply janusgraph-driver/src/main/java/org/janusgraph/graphdb/relations/RelationIdentifier.java:63]: Thanks for the explanation! I guess it makes sense to just return the `inVertexId` as is. If it's not initialized, e.g., for a property, and therefore will be `null`, then the caller has to decide how to handle that.

**Comment by li-boxuan** [comment]
I'll merge this PR at the end of this week if no one else plans to review it.

**Review by porunov — COMMENTED** [review]
I see many places where we still use `longId` assuming a vertex has `long` id and not custom id (for example, below in the comment). Will it work normally? How will JanusGraph messaging bus work with non-long vertex ids? I assume it may be a breaking change because logs will now have non-long ids, but I could be wrong.

I also see `IndexRecordUtil.element2String(Object elementId)` which assumes that the id is either `Long` or `RelationIdentifier` and never anything else.

You could potentially also change signature of the next method: https://github.com/JanusGraph/janusgraph/blob/6549949a9c734da19527ce4808e6be51c53f6a68/janusgraph-core/src/main/java/org/janusgraph/graphdb/database/util/StaleIndexRecordUtil.java#L164 but it's not necessary. I can change it later after this PR is merged.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/computer/VertexMemoryHandler.java
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/computer/VertexMemoryHandler.java:157]: This PR does not intend to bring about any breaking or functional change. For now, we just assume its a long value, but in the subsequent PRs, this will be changed. The message bus will also require breaking changes.

Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/computer/VertexMemoryHandler.java:157]: btw

>  You could potentially also change signature of the next method

`forceRemoveVertexFromGraphIndex` is changed in this PR. Maybe you are looking at the wrong version?

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/computer/VertexMemoryHandler.java:157]: You right. I missed that. Thank you!

File: janusgraph-core/src/main/java/org/janusgraph/util/datastructures/AbstractIdListUtil.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/util/datastructures/AbstractIdListUtil.java:48]: Why `Double` and not `Long`? I don't remember but aren't we loosing precision after 2^53 here?

File: janusgraph-driver/src/main/java/org/janusgraph/util/IDUtils.java
Comment (porunov) [inline janusgraph-driver/src/main/java/org/janusgraph/util/IDUtils.java:28]: Not related to this PR but do you know by any chance why negative vertex ids are not allowed?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/ElementUtils.java
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/ElementUtils.java:42]: I combined `id instanceof Vertex` case with `id instanceof JanusGraphVertex`. See line 29 of new version.

File: janusgraph-driver/src/main/java/org/janusgraph/util/IDUtils.java
Comment (li-boxuan) [reply janusgraph-driver/src/main/java/org/janusgraph/util/IDUtils.java:28]: Iff `ids.flush=false`, then JanusGraph will assign a temporary negative ID when an element is created and generates a unique positive ID only when the transaction commits. Thus, a negative number represents a temporary id.

Comment (porunov) [reply janusgraph-driver/src/main/java/org/janusgraph/util/IDUtils.java:28]: Thank you, got it! I guess it would make sense to think about reducing that scope (I guess `2^31` would be enough for temporary ids and the rest could be potentially used for normal ids) but it's definitely out of the scope of this PR.

File: janusgraph-core/src/main/java/org/janusgraph/util/datastructures/AbstractIdListUtil.java
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/util/datastructures/AbstractIdListUtil.java:48]: Great catch! And I shouldn't have allowed double value in the first place.

**Review by porunov — APPROVED** [review]
LGTM! Thank you @li-boxuan !

**Comment by m-thirumal** [comment]
This  feature will be available for release 1.0.0? Because this is not available in Release candidate `janusgraph-1.0.0-rc1`. 

Still getting the same error

```
gremlin> g.addV('user').property(T.id, 'f3695384-8442-447e-ba25-914e5b8deb08').property('name', 'Thirumal')
Cannot cast java.lang.String to java.lang.Number
```

**Comment by li-boxuan** [comment]
@m-thirumal This feature is WIP. I have the code ready but need some time to clean it up and let the community review it. It may or may not be released in 1.0.0.

**Pull request merged**
